modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeleteSafeZoneCalc.java

```diff
@@ -0,0 +1,206 @@
+    }
+
+    @Override
+    public void run() {
+        if (log.isDebugEnabled()) {
+            log.debug("Slot deletion safe zone calculation started.");
+        }
+        while (running) {
+            if (isLive) {
+                Set<String> nodesWithPublishedMessages = SlotManager.getInstance().getMessagePublishedNodes();
+                Map<String, Long> nodeInformedSafeZones =
+                        SlotManager.getInstance().getNodeInformedSlotDeletionSafeZones();
+
+                /** calculate safe zone (minimum value of messageIDs published so far to the
+                 * cluster by each node)
+                 */
+                long globalSafeZoneVal = Long.MAX_VALUE;
+                for (String nodeID : nodesWithPublishedMessages) {
+
+                    long safeZoneValue = Long.MAX_VALUE;
+
+                    //get the maximum message id published by node so far
+                    Long safeZoneByPublishedMessages = SlotManager.getInstance()
+                            .getLastPublishedIDByNode(nodeID);
+
+                    if (null != safeZoneByPublishedMessages) {
+                        safeZoneValue = safeZoneByPublishedMessages;
+                    }
+
+                    //If messages are not published, each node will send a messageID giving
+                    // assurance that next message id it would generate will be beyond a certain
+                    // number
+                    Long nodeInformedSafeZone = nodeInformedSafeZones.get(nodeID);
+
+                    /**
+                     * if no new messages are published and no new slot assignment happened
+                     * node informed value can be bigger. We need to accept that to keep the
+                     * safe zone moving
+                     */
+                    if (null != nodeInformedSafeZone) {
+                        if (Long.MAX_VALUE != safeZoneValue) {
+                            if (safeZoneValue < nodeInformedSafeZone) {
+                                safeZoneValue = nodeInformedSafeZone;
+                            }
+                        } else {
+                            safeZoneValue = nodeInformedSafeZone;
+                        }
+                    }
+
+                    if (globalSafeZoneVal > safeZoneValue) {
+                        globalSafeZoneVal = safeZoneValue;
+                    }
+                }
+
+                slotDeleteSafeZone = globalSafeZoneVal;
+
+                if (log.isDebugEnabled()) {
+                    log.info("Safe Zone Calculated : " + slotDeleteSafeZone);
+                }
+
+                try {
+                    Thread.sleep(seekInterval);
+                } catch (InterruptedException e) {
+                    //silently ignore
+                }
+            } else {
+                try {
+                    Thread.sleep(15 * 1000);
+                } catch (InterruptedException e) {
+                    //silently ignore
```

Comment:
Do we need to ignore a thread interrupt here? Can't we have something like
`thread.currentthread().interrupt()`

Anyway can't we have this whole thing as a sheduled task. Do we need to go in an infinite loop updating the safe zone and sleeping. This is not in critical path right? Plus wouldn't it make this Thread.sleep() stuff go away?